uint8 next_state
